Expose a Python interface for inference functions

KubinGH opened the pull request:
Exposes a node-level API for performing type/shape inference to Python bindings.

**Description**
(Updated)

- Adds a low-level function `OpSchema._infer_node_outputs`, to which serialised protobuf bytes are passed - arguments: schema, node, input types, input data/other inference context arguments, returns: inferred output types. 
- Adds a public interface `shape_inference.infer_node_outputs`:
```py
def infer_node_outputs(
     schema: onnx.defs.OpSchema,
     node: onnx.NodeProto,
     input_types: Dict[str, onnx.TypeProto],
     input_data: Optional[Dict[str, onnx.TensorProto]] = None,
     input_sparse_data: Optional[Dict[str, onnx.SparseTensorProto]] = None,
 ) -> Dict[str, onnx.TypeProto]:  # output_types
```
  - Underneath this performs validation of passed in nodes/types, `ValidationError`s and `InferenceError`s are raised.
  - Should be the only used function, `OpSchema._infer_node_outputs` is internal to the implementation.
  - `input_types` keys are named like `node` declares its inputs.
  - `input_data` and `input_sparse_data` are used for passing in known constant inputs (as per C++ usage).
  - Could implement passing in partial data propagation results and `GraphInferenceContext` in the future.
- Adds test cases for some inference calls from the Python side.

**Motivation and Context**
- See #4405

KubinGH commented:
@gramalingam Would you mind viewing the proposal if this sort of interface would work? In the future it could be extended to support the other arguments, they could be optional so it's backward compatible.

gramalingam commented on onnx/defs/schema.h:
```diff
@@ -818,6 +818,10 @@ class OpSchema final {
     return tensor_inference_function_ ? true : false;
   }
 
+  InferenceFunction get_type_and_shape_inference_function() const {
```
I think this getter function already exists: see `GetTypeAndShapeInferenceFunction`.

KubinGH replied:
You're right, I think I must have thought it was private for some reason. Fixed!

gramalingam commented:
Updating the pyi definition for [OpSchema](https://github.com/onnx/onnx/blob/main/onnx/onnx_cpp2py_export/defs.pyi#L7) would be useful.

gramalingam commented on onnx/test/inference_function_test.py:
```diff
@@ -0,0 +1,64 @@
+import onnx.shape_inference
+
+
+def call_infer_types(
```
I think it might be useful to promote an API like this as a wrapper around the raw C++ method (that is, make if part of the public API instead of something in just this test-file). But, for that, I don't think we would want `node` to be optional. And am not sure about `num_outputs` either.

KubinGH replied:
Sure, that definitely makes sense. We could always require a `NodeProto`, `num_outputs` was a temporary idea.

gramalingam commented on onnx/cpp2py_export.cc:
```diff
@@ -39,6 +39,47 @@ static std::string ProtoBytesToText(const py::bytes& bytes) {
+  return make_pair(move(values), result);
+}
+
+std::vector<py::bytes> CallInferenceFunction(
```
nit: for various functions it might be useful to explicitly indicate it is a node-level inference (vs. graph-level inference) in the name, eg: `CallNodeInferenceFunction` and `infer_types_for_node`.

KubinGH replied:
That's a fair idea, I'll adapt to this in the next iteration.

KubinGH commented:
> I think it might be useful to promote an API like this as a wrapper around the raw C++ method (that is, make if part of the public API instead of something in just this test-file). But, for that, I don't think we would want `node` to be optional. And am not sure about `num_outputs` either.

When it comes to the Python-side interface: A node has input/output names, operator name & domain, attributes. Own name and docstring don't matter. We are missing basically only the operator version it seems, and the input types. I see some variations of the possible signature:

- Inputs & outputs
  - `list[TypeProto]` - we already have an ordering of fields inside the node. Node inputs/outputs are positional anyway, so making it explicit with ordering in a `list` seems okay.
  - `dict[str, TypeProto]` - since the node names the fields in definition. On the other hand since the node has to get created explicitly first anyway in this approach we might as well use the names that are within the node. I think this shouldn't mess with generics etc. since if we use a name multiple times is has to have the same type - we get some very simple consistency check.
- Operator version
  - `version: int` - get the current operator schema from a list at this current version based on what is in the node.
  - `OpSchema`- make the user explicitly provide the schema themselves. Is more redundant, since the Node already has the domain and identifier defined. 


Also, where would we put this function? `shape_inference` would be fine, but really it should have been called `inference` (as it also infers types and other stuff?). Could also put it in `helper`. Any ideas?

Let me know your thoughts and I'll get started on some improvements in the Python-interface area.

KubinGH commented:
@gramalingam I pushed some changes:

```cpp
std::unordered_map<std::string, py::bytes> CallNodeInferenceFunction(
    OpSchema* schema,
    const py::bytes& nodeBytes,
    std::unordered_map<std::string, py::bytes> valueTypesByNameBytes,
    std::unordered_map<std::string, py::bytes> inputDataByNameBytes,
    std::unordered_map<std::string, py::bytes> inputSparseDataByNameBytes) {
```

- Now returns a map, since the inputs are a map already. Doesn't return entries were the output `TypeProto` was not initialised (inference failed implicitly?).
- Added `OpSchema::Verify` before the call to the inference (checks if the count of inputs/outputs is correct, etc.), and `OpSchema::CheckInputOutputType` afterwards (checks if the types are consistent with the signature). Both should throw `ValidationError`.
- At the same time, inference itself may throw `InferenceError`. Interestingly, this may also throw errors related to attributes (so things like `Cast` will check here if exactly one of the right attributes is set).
- Let me know if I should handle exceptions there in some more involved way. I think what is thrown already reflects most of what `check_model` and `infer_shapes` do.

```py
# class OpSchema:
    def infer_node_outputs(self, node_proto: bytes, value_types: Dict[str, bytes],
                           input_data: Dict[str, bytes], input_sparse_data: Dict[str, bytes]
                           ) -> Dict[str, bytes]: ...
```

- This is just a binding for the C++ function. I added it to `defs.pyi` as above. 
- My general feeling is that the attributes after `value_types` should have empty defaults so that in the future `generatedShapeData` and `graphInferenceContext` may be supported as well (but this involves, I think, significant work).

```py
# shape_inference.py
def infer_node_outputs(
    schema: onnx.defs.OpSchema, node: onnx.NodeProto, input_types: Dict[str, onnx.TypeProto],
    input_data: Optional[Dict[str, onnx.TensorProto]] = None,
    input_sparse_data: Optional[Dict[str, onnx.SparseTensorProto]] = None
) -> Dict[str, onnx.TypeProto]: ...
```

- The public-facing Python function does all of the conversions and expects proper types.
- Already has default arguments, which we may change in the future. Reflect the actual underlying implementation.
- Does some checks if all the expected keys are present and only copies over what is actually required on C++ side for some performance.


Let me know what you think of these changes! I also added some more tests.

gramalingam commented on onnx/test/inference_function_test.py:
```diff
@@ -1,63 +1,73 @@
-            onnx.helper.make_tensor_type_proto(1, (1, 2))
-        ], 1) == [onnx.helper.make_tensor_type_proto(1, (None, 2))]
+        cases = [
+            ({'A': (1, ()), 'B': (1, ())}, {'C': (1, ())}),
```
nit: I assume the 1 is the type encoding of float. If so, better to use the symbolic names like `TensorProto.FLOAT` instead.

KubinGH replied:
I'll fix this (and the Windows CI, looks like it's failing due to a different default int dtype for a numpy array?) tomorrow.

gramalingam replied:
I heard that the Windows CI are failing across PRs, may be unrelated to this PR. Thanks!

gramalingam commented on onnx/cpp2py_export.cc:
```diff
@@ -39,6 +39,57 @@ static std::string ProtoBytesToText(const py::bytes& bytes) {
+  // Construct inference context and get results - may throw InferenceError
+  shape_inference::InferenceContextImpl ctx(node, valueTypes.second, inputData.second, inputSparseData.second);
+  schema->GetTypeAndShapeInferenceFunction()(ctx);
+  // Verify the inference succeeded - may also throw ValidationError
```
A question: wondering what happens in builds that don't support exceptions. I assume that the C++ code with abort. What does pybind return in this case? (I realize that this is a pre-existing issue, but just wondering whether we should mark/exclude tests in some special way for this case.)

KubinGH replied:
Looks like this isn't really going to be an issue, since `cmake_build` under Python requires the bindings to be built with exceptions:

https://github.com/onnx/onnx/blob/f6cb158986cf65feae64774d9a851e2649b94073/setup.py#L207-L208

xadupre commented on onnx/cpp2py_export.cc:
```diff
@@ -85,6 +136,10 @@ PYBIND11_MODULE(onnx_cpp2py_export, onnx_cpp2py_export) {
       .def_property_readonly("has_context_dependent_function", &OpSchema::HasContextDependentFunction)
+      .def("infer_node_outputs", CallNodeInferenceFunction,
+          py::arg("nodeBytes"), py::arg("valueTypesByNameBytes"),
+          py::arg("inputDataByNameBytes") = std::unordered_map<std::string, py::bytes>{},
```
None is better than an empty container as a default value.

JakubBachurskiQC replied:
What do you mean by None here? A null pointer? 
I think that not having any information in the map makes sense as a default for inference.

KubinGH replied:
I don't think I know what you mean here. This is C++ side, so there's no None. Do you mean to make it a pointer and use nullptr/use std::optional? 
I think that this wouldn't be a lot better as the map isn't mutated and also in this it makes sense that the default is "no additional information" (for the inference context we have to pass in _some_ map, and it makes sense to make it empty as there's no input data we know).
Could you provide a code example for the change?

gramalingam reviewed (approved):
Looks great, thanks very much!

KubinGH commented:
Thanks! 
Should be ready for merge, then?

justinchuby commented on onnx/shape_inference.py:
```diff
@@ -82,4 +82,42 @@ def infer_shapes_path(
+def infer_node_outputs(
+    schema: onnx.defs.OpSchema,
+    node: onnx.NodeProto,
+    input_types: Dict[str, onnx.TypeProto],
```
Do we explicitly require a dictionary for these params? If not prefer a more generic type like Mapping. 

——-

Use more generic types for input parameters and specific types for return values. For example, a function may take a `Sequence` and return a `List`.

Diagram for reference: https://gist.github.com/justinchuby/4021cebe9e093f636759a88de325c85f
